### PR TITLE
WIP: Work to complete the LLVM graph operations

### PR DESCRIPTION
**Pull request by mariari**

### CODE REVIEW

File: src/Juvix/Backends/LLVM/Codegen/Graph.hs [hunk 1]
```diff
@@ -29,46 +33,58 @@ link ∷
...
+    HasState "count" Word m,
+    HasState "currentBlock" Name.Name m,
+    HasState "moduleDefinitions" [AST.Definition] m,
+    HasState "names" Names m,
+    HasState "symtab" (Map.HashMap Symbol Operand.Operand) m,
+    HasState "typTab" TypeTable m,
+    HasState "varTab" VariantToType m
+  ) ⇒
+  m Operand.Operand
+isBothPrimary = Block.defineFunction Type.i1 "is_both_primary" args $
```
Comment (cwgoes) [inline src/Juvix/Backends/LLVM/Codegen/Graph.hs]: What is this function doing? Checking if two ports are a primary pair?

File: src/Juvix/Backends/LLVM/Codegen/Graph.hs [hunk 2]
```diff
@@ -29,46 +33,58 @@ link ∷
...
+  do
+    -- TODO ∷ should this call be abstracted somewhere?!
+    -- Why should Ι allocate for every port?!
+    mainPort ← allocaNumPortsStatic False (Operand.ConstantOperand (C.Int 32 0))
```
Comment (cwgoes) [inline src/Juvix/Backends/LLVM/Codegen/Graph.hs]: Does this just declare a local variable?

File: src/Juvix/Backends/LLVM/Net/EAC.hs [hunk 3]
```diff
@@ -0,0 +1,57 @@
...
+-- Reduction rules
+--------------------------------------------------------------------------------
+
+-- TODO ∷ Maybe add metadata at some point?
+
+-- These functions work off the nodeType signature not Types.eac
+
+-- mimic rules from the interpreter
+-- This rule applies to Application ↔ Lambda
+anihilateRewireAux = Codegen.defineFunction Type.void "anihilate_rewire_aux" args $
```
Comment (cwgoes) [inline src/Juvix/Backends/LLVM/Net/EAC.hs]:
```suggestion
annihilateRewireAux = Codegen.defineFunction Type.void "annihilate_rewire_aux" args $
```

File: src/Juvix/Backends/LLVM/Net/EAC.hs [hunk 4]
```diff
@@ -0,0 +1,57 @@
...
+
+--------------------------------------------------------------------------------
+-- Reduction rules
+--------------------------------------------------------------------------------
+
+-- TODO ∷ Maybe add metadata at some point?
+
+-- These functions work off the nodeType signature not Types.eac
+
+-- mimic rules from the interpreter
```
Comment (cwgoes) [inline src/Juvix/Backends/LLVM/Net/EAC.hs]: I wonder if we can split these out neatly into constructor/duplicator rules. If not, it's OK, but it'd be nice.

File: src/Juvix/Backends/LLVM/Codegen/Graph.hs [hunk 5]
```diff
@@ -29,46 +33,58 @@ link ∷
...
+    -- TODO ∷ Make sure findEdge is in the environment
+    edge ← Block.externf "find_edge"
+    nodePtr ← Block.externf "node_ptr"
+    node ← load nodeType nodePtr
```
Comment (cwgoes) [inline src/Juvix/Backends/LLVM/Codegen/Graph.hs]: Are we loading the node data here? Do we need to do that to read the port address?

File: src/Juvix/Backends/LLVM/Codegen/Graph.hs [hunk 6]
```diff
@@ -29,46 +33,58 @@ link ∷
...
+    port ← call portType edge (Block.emptyArgs [node, mainPort])
+    otherNodePtr ← getElementPtr $
+      Types.Minimal
+        { Types.type' = nodePointer,
+          Types.address' = port,
+          Types.indincies' = Block.constant32List [0, 1]
+        }
+    -- convert ptrs to ints
```
Comment (cwgoes) [inline src/Juvix/Backends/LLVM/Codegen/Graph.hs]: Hopefully this is a zero-cost cast under the hood.

File: src/Juvix/Backends/LLVM/Codegen/Graph.hs (hunk 1 above)
Comment (mariari) [reply src/Juvix/Backends/LLVM/Codegen/Graph.hs]: This checks if two ports are connected via their primary pair

File: src/Juvix/Backends/LLVM/Codegen/Graph.hs (hunk 2 above)
Comment (mariari) [reply src/Juvix/Backends/LLVM/Codegen/Graph.hs]: This allocs the main port, which is the 0th index of the port array

File: src/Juvix/Backends/LLVM/Codegen/Graph.hs (hunk 6 above)
Comment (mariari) [reply src/Juvix/Backends/LLVM/Codegen/Graph.hs]: If the pointer is the same size as the int here, it's a 0 cost cast!

File: src/Juvix/Backends/LLVM/Codegen/Graph.hs (hunk 5 above)
Comment (mariari) [reply src/Juvix/Backends/LLVM/Codegen/Graph.hs]: we need to load the node, as find_edge requires a nodeType and not a nodePointer

File: src/Juvix/Backends/LLVM/Net/EAC.hs (hunk 4 above)
Comment (mariari) [reply src/Juvix/Backends/LLVM/Net/EAC.hs]: the rules from the original interpreter?

Comment (cwgoes) [reply src/Juvix/Backends/LLVM/Net/EAC.hs]: Not exactly, just sharing common logic where applicable.

**Comment by cwgoes** [comment]
Merged, follow-up later (ref https://github.com/cryptiumlabs/juvix/issues/189)

**Pull request merged**